In [114]:
import pandas as pd
import numpy as np

In [115]:
data=pd.read_csv('/content/Fraud.csv')

In [86]:
data.head(10)

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815      170136.00       160296.36   
1     1   PAYMENT   1864.28  C1666544295       21249.00        19384.72   
2     1  TRANSFER    181.00  C1305486145         181.00            0.00   
3     1  CASH_OUT    181.00   C840083671         181.00            0.00   
4     1   PAYMENT  11668.14  C2048537720       41554.00        29885.86   
5     1   PAYMENT   7817.71    C90045638       53860.00        46042.29   
6     1   PAYMENT   7107.77   C154988899      183195.00       176087.23   
7     1   PAYMENT   7861.64  C1912850431      176087.23       168225.59   
8     1   PAYMENT   4024.36  C1265012928        2671.00            0.00   
9     1     DEBIT   5337.77   C712410124       41720.00        36382.23   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0            0.00      0.0             0.0  
1  M2044282225             0.0            0.00      0.0             0.0  
2   C553264065             0.0            0.00      1.0             0.0  
3    C38997010         21182.0            0.00      1.0             0.0  
4  M1230701703             0.0            0.00      0.0             0.0  
5   M573487274             0.0            0.00      0.0             0.0  
6   M408069119             0.0            0.00      0.0             0.0  
7   M633326333             0.0            0.00      0.0             0.0  
8  M1176932104             0.0            0.00      0.0             0.0  
9   C195600860         41898.0        40348.79      0.0             0.0

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206154 entries, 0 to 206153
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   step            206154 non-null  int64  
 1   type            206154 non-null  object 
 2   amount          206154 non-null  float64
 3   nameOrig        206154 non-null  object 
 4   oldbalanceOrg   206154 non-null  float64
 5   newbalanceOrig  206154 non-null  float64
 6   nameDest        206154 non-null  object 
 7   oldbalanceDest  206153 non-null  float64
 8   newbalanceDest  206153 non-null  float64
 9   isFraud         206153 non-null  float64
 10  isFlaggedFraud  206153 non-null  float64
dtypes: float64(7), int64(1), object(3)
memory usage: 17.3+ MB


In [94]:
data.isna()

step   type  amount  nameOrig  oldbalanceOrg  newbalanceOrig  \
0        False  False   False     False          False           False   
1        False  False   False     False          False           False   
2        False  False   False     False          False           False   
3        False  False   False     False          False           False   
4        False  False   False     False          False           False   
...        ...    ...     ...       ...            ...             ...   
5098279  False  False   False     False          False           False   
5098280  False  False   False     False          False           False   
5098281  False  False   False     False          False           False   
5098282  False  False   False     False          False           False   
5098283  False  False   False     False          False           False   

         nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0           False           False           False    False           False  
1           False           False           False    False           False  
2           False           False           False    False           False  
3           False           False           False    False           False  
4           False           False           False    False           False  
...           ...             ...             ...      ...             ...  
5098279     False           False           False    False           False  
5098280     False           False           False    False           False  
5098281     False           False           False    False           False  
5098282     False           False           False    False           False  
5098283     False           False            True     True            True  

[5098284 rows x 11 columns]

In [95]:
data.isna().sum()

step              0
type              0
amount            0
nameOrig          0
oldbalanceOrg     0
newbalanceOrig    0
nameDest          0
oldbalanceDest    0
newbalanceDest    1
isFraud           1
isFlaggedFraud    1
dtype: int64

**DATA CLEANING**


*   since there are null values in the dataset and their occurance is only once, thus we can safely remove that row (as deleting one row won't affect the dataset containing over 5 million rows)





In [116]:
df=data.dropna() #removing the null values

In [117]:
df.isna().sum()

step              0
type              0
amount            0
nameOrig          0
oldbalanceOrg     0
newbalanceOrig    0
nameDest          0
oldbalanceDest    0
newbalanceDest    0
isFraud           0
isFlaggedFraud    0
dtype: int64

The sender's and receiver's id's are of no use as they don't really contribute in prediction. Thus they are removed from the dataset and new dataset is stored in df variable. Same goes for type of transacion, i tried applying hot label encoding to include it, but it did not gave any better results so i just removed it.

In [118]:
df=df.drop(['nameOrig','nameDest','type'],axis=1)

In [107]:
df

step     amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
0           1    9839.64      170136.00       160296.36            0.00   
1           1    1864.28       21249.00        19384.72            0.00   
2           1     181.00         181.00            0.00            0.00   
3           1     181.00         181.00            0.00        21182.00   
4           1   11668.14       41554.00        29885.86            0.00   
...       ...        ...            ...             ...             ...   
5098278   355  242198.32       19118.00       261316.32      1516040.70   
5098279   355  913176.92      261316.32            0.00      1825236.29   
5098280   355  201426.05           0.00            0.00      4763749.53   
5098281   355  190513.87           0.00            0.00      2404718.74   
5098282   355  352036.31           0.00            0.00      2260267.83   

         newbalanceDest  isFraud  isFlaggedFraud  
0                  0.00      0.0             0.0  
1                  0.00      0.0             0.0  
2                  0.00      1.0             0.0  
3                  0.00      1.0             0.0  
4                  0.00      0.0             0.0  
...                 ...      ...             ...  
5098278      1273842.38      0.0             0.0  
5098279      2973636.63      0.0             0.0  
5098280      4965175.58      0.0             0.0  
5098281      2737963.54      0.0             0.0  
5098282      2612304.14      0.0             0.0  

[5098283 rows x 8 columns]

Now we're assigning our X and Y variables with the values from the dataset.
X includes all the columns if df except isFraud as it our dependent variable.

In [119]:
X=df.loc[:,df.columns!='isFraud']
Y=df['isFraud']

In [109]:
X

step     amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
0           1    9839.64      170136.00       160296.36            0.00   
1           1    1864.28       21249.00        19384.72            0.00   
2           1     181.00         181.00            0.00            0.00   
3           1     181.00         181.00            0.00        21182.00   
4           1   11668.14       41554.00        29885.86            0.00   
...       ...        ...            ...             ...             ...   
5098278   355  242198.32       19118.00       261316.32      1516040.70   
5098279   355  913176.92      261316.32            0.00      1825236.29   
5098280   355  201426.05           0.00            0.00      4763749.53   
5098281   355  190513.87           0.00            0.00      2404718.74   
5098282   355  352036.31           0.00            0.00      2260267.83   

         newbalanceDest  isFlaggedFraud  
0                  0.00             0.0  
1                  0.00             0.0  
2                  0.00             0.0  
3                  0.00             0.0  
4                  0.00             0.0  
...                 ...             ...  
5098278      1273842.38             0.0  
5098279      2973636.63             0.0  
5098280      4965175.58             0.0  
5098281      2737963.54             0.0  
5098282      2612304.14             0.0  

[5098283 rows x 7 columns]

Now we're dividing the dataset into training and testing data

In [124]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.40,random_state=10)

Finally i applied LogisticRegression algorithm (as the x and y graph tend to be in a straight line (linear regression)).

In [126]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression()
clf.fit(X_train, Y_train)
Y_pred = clf.predict(X_test)
print ("Accuracy : ", clf.score(X_test, Y_test))

Accuracy :  0.998324196635977


**The model was finally able to detect the fraud cases with an accuracy of 99.83 %**

**Here are the answers to some FAQs**

1. Data cleaning including missing values, outliers and multi-collinearity.


* As i explained above since there are null values in the dataset and theiroccurance is only once, thus we can safely remove that row (as deleting one row won't affect the dataset containing over 5 million rows)




2. Describe your fraud detection model in elaboration.


*   After applying all the data cleaning and filler column removal stuff, the dataset was divided into training and testing data after assigning columns to X and Y respectively

* After that the model was trained using LogisticRegression algorithm,as the data points formed a linear regression. The dependent variable Y was analysed over different sets of X.

* Finally the trained model was tested on testing data which predicted the results with an accuracy of 99.83 %



3. How did you select variables to be included in the model?


*   X includes all the columns if df except isFraud as it our dependent variable. It analyzed the columns like oldbalanceDest , newbalanceDest ,amount etc. and saw their relationship with isFraud
*   Y was assigned with isFraud column.






4. Demonstrate the performance of the model by using best set of tools.


*   The naive bayes algo predicted the results with 97% accuracy
*   Since the datapoints were linear in nature thus i applied LogisticRegression which predicted the results with 99.83% accuracy



5. What are the key factors that predict fraudulent customer?


*   The amount transferred and oldbalanceDest, newbalanceDest are majorly responsible for fraud detection as in case of fraud the oldbalanceDest and newbalanceDest diffrence don't match with amount




6. Do these factors make sense? If yes, How? If not, How not?


*   Yes, the factors which i mentioned do make sense



7. What kind of prevention should be adopted while company update its infrastructure?

*   The company should insert a strong fraud detection Ml model in their infrastructure



8. Assuming these actions have been implemented, how would you determine if they work?

*  we can ask the company to let them insert some dummy transactions for testing purpose

